internals/src/error/input_string.rs

```diff
@@ -0,0 +1,110 @@
+//! Implements the [`InputString`] type storing the parsed input.
+
+use storage::Storage;
+use core::fmt;
+
+/// Conditionally stores the input string in parse errors.
+///
+/// This type stores the input string of a parse function depending on whether `alloc` feature is
+/// enabled. When it is enabled, the string is store inside as `String`. When disabled this is a
+/// zero-sized type and attempt to store a string does nothing.
+///
+/// This provides two methods to format the error strings depending on the context.
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct InputString(Storage);
+
+impl InputString {
+    /// Displays a message saying "failed to parse <self> as <what>".
+    ///
+    /// This is normally used whith the `write_err!` macro.
+    pub fn display_cannot_parse<'a, What: fmt::Display + ?Sized>(&'a self, what: &'a What) -> CannotParse<'a, What> {
+        CannotParse {
+            input: self,
+            what,
+        }
+    }
+
+    /// Formats a message saying "<self> is not a known <what>".
+    ///
+    /// This is normally used in leaf parse errors (with no source) when parsing an enum.
+    pub fn unknown_variant<What: fmt::Display + ?Sized>(&self, what: &What, f: &mut fmt::Formatter) -> fmt::Result {
+        storage::unknown_variant(&self.0, what, f)
+    }
+}
+
+macro_rules! impl_from {
+    ($($type:ty),+ $(,)?) => {
+        $(
+            impl From<$type> for InputString {
+                fn from(input: $type) -> Self {
+                    #[allow(clippy::useless_conversion)]
+                    InputString(input.into())
+                }
+            }
+        )+
+    }
+}
+
+impl_from!(&str);
+
+/// Displays message saying "failed to parse <input> as <what>".
+///
+/// This is created by `display_cannot_parse` method and should be used as
+/// `write_err!("{}", self.input.display_cannot_parse("what is parsed"); self.source)` in parse
+/// error [`Display`](fmt::Display) imlementation.
```

Comment:
I had a go at using this PR on top of the new `bitcoin-units` crate, I think this should be:

```suggestion
/// This is created by `display_cannot_parse` method and should be used as
/// `write!("{}", self.input.display_cannot_parse("what is parsed"))` in parse
/// error [`Display`](fmt::Display) imlementation.
```

Reply:
What? I thought `InputString` replaces a `String` in an error variant, how can it have another source?

Reply:
FTR this is a snippet of how I used it:
```rust
            UnknownDenomination(ref d) =>
                write!(f, "unknown denomination: {}", d.display_cannot_parse("a bitcoin denomination")),
            PossiblyConfusingDenomination(ref d) => {
                write!(f, "the first character of denomination has a technical meaning but that denomination is uncommon so it is possibly confusing: {}", d.display_cannot_parse("a bitcoin denomination"))
```

Reply:
No worries, I can't think of a clear example,  I rekon it would be better to include the most simple case in this doc but I won't hold the PR up for this.

Reply:
> how can it have another source?

It doesn't. The error it's in may have. E.g. our `ParseIntError` has both input string and `core::num::ParseIntError` which is the source.

Reply:
Yeah that's what I meant, in my mind this is an unusual case. Much more likely is
```rust
pub enum FooError {
    Bar,
    Parse(InputString),
}
```

But its no biggy.

Reply:
Oh there is literally a macro that defines a type like this :)